### PR TITLE
KAFKA-6106; Postpone normal processing of tasks within a thread until…

### PR DESCRIPTION
**Pull request by kamalcph**
… restoration of all tasks have completed.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by kamalcph** [comment]
@guozhangwang @mjsax 

Call for review. Should I have to clear the `resumedPartitions` when `onPartitionsRevoked` triggers ? (Not sure)

**Comment by mjsax** [comment]
Thanks for the patch, @kamalcph 

I am not sure if this is sufficient to resolve the ticket. If I am not wrong, we only pause partitions for newly assigned tasks (cf `TaskManager#createTasks` L112-114). However, on restore we should pause all partitions during restore. Does this make sense?

**Comment by mjsax** [comment]
\cc @bbejeck

**Comment by kamalcph** [comment]
@mjsax Thanks for the pointer. Reassigned and new tasks with no state stores are not being paused. One option is to pause all the assigned partitions in `TaskManager#createTasks` until the stores are ready but some `StreamThreadTest` are failing. I'm working on it.

**Comment by guozhangwang** [comment]
+1 to @mjsax 's comment.

@kamalcph I'd suggest you also consider do some code refactoring as we would only resume consumer on all partitions once all tasks have completed restoration. For example, `consumer.resume(resumed);` should be extracted outside of `updateNewAndRestoringTasks`.

**Review by mjsax — COMMENTED** [review]
I guess this is still not sufficient. We need to make sure to pause all partitions. Atm, we only pause a partitions if its task has a state store.

cf. `TaskManager#createTasks` and `AssignedTasks#uninitializedPartitions`

I think we can remove `AssignedTasks#uninitializedPartitions` and pass to complete `assignment` that is passed into `TaskManager#createTasks` as parameter

**Comment by mjsax** [comment]
Btw: please leave a comment after you update the PR. Otherwise, we might miss that you did update it.

**Comment by kamalcph** [comment]
@mjsax, Sorry for the delay.  Last time, when I updated the code, I ran only the TaskManagerTest. After push, I realized that some of the StreamThreadTest cases are failing. So, I reverted the code immediately (this may caused the update in PR).

 I’ve updated this PR now. Still facing this exception in StreamThread test cases. 

    java.lang.IllegalStateException: No current assignment for partition topic1-1
	at org.apache.kafka.clients.consumer.internals.SubscriptionState.assignedState(SubscriptionState.java:259)
	at org.apache.kafka.clients.consumer.internals.SubscriptionState.pause(SubscriptionState.java:413)
	at org.apache.kafka.clients.consumer.MockConsumer.pause(MockConsumer.java:341)
	at org.apache.kafka.streams.processor.internals.TaskManager.createTasks(TaskManager.java:114)
	at org.apache.kafka.streams.processor.internals.StreamThreadTest.shouldReturnActiveTaskMetadataWhileRunningState(StreamThreadTest.java:717)

Previously, the partitions returned by `AssignedTasks#uninitializedPartitions` was empty. So, this error never occurred. I tried to subscribe (topic1) in mock consumer but it didn’t succeeded. Could you please help me out ?

**Comment by mjsax** [comment]
I see. Thanks.

About the test. I am not sure, but maybe you need to call `thread.rebalanceListener.onPartitionsAssigned(assignedPartitions);` before `thread.taskManager().createTasks(assignedPartitions);`

**Comment by kamalcph** [comment]
@mjsax Test cases are passed now. Please review it.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 1]
```diff
@@ -323,18 +323,17 @@ void setConsumer(final Consumer<byte[], byte[]> consumer) {
      * @throws TaskMigratedException if another thread wrote to the changelog topic that is currently restored
      */
     boolean updateNewAndRestoringTasks() {
-        final Set<TopicPartition> resumed = active.initializeNewTasks();
+        active.initializeNewTasks();
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:325]: I think we can cleanup `initializeNewTasks()` internally and change its return type to `void` ?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 2]
```diff
@@ -323,18 +323,17 @@ void setConsumer(final Consumer<byte[], byte[]> consumer) {
...
         standby.initializeNewTasks();
 
         final Collection<TopicPartition> restored = changelogReader.restore(active);
 
-        resumed.addAll(active.updateRestored(restored));
+        active.updateRestored(restored);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:330]: Same for `updateRestored()`

**Review by mjsax — COMMENTED** [review]
Some more comments. Did not look at the tests in detail yet.

Call for second review @guozhangwang @bbejeck @dguy @vvcephei

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 3]
```diff
@@ -109,9 +109,9 @@ void createTasks(final Collection<TopicPartition> assignment) {
         active.closeNonAssignedSuspendedTasks(assignedActiveTasks);
         addStreamTasks(assignment);
         addStandbyTasks();
-        final Set<TopicPartition> partitions = active.uninitializedPartitions();
-        log.trace("Pausing partitions: {}", partitions);
-        consumer.pause(partitions);
+        // Pause all the partitions until the underlying state store is ready for all the active tasks.
+        log.trace("Pausing partitions: {}", assignment);
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:113]: I'm wondering if this should be `debug` or not, same below when resuming assignment, although I don't feel too strongly about this.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 4]
```diff
@@ -626,13 +628,19 @@ public void shouldPunctuateActiveTasks() {
     }
 
     @Test
-    public void shouldResumeConsumptionOfInitializedPartitions() {
+    public void shouldNotResumeConsumptionUntilAllStoresRestored() {
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:631]: For this test would it be more clear to define a local strict mock for the consumer `Consumer<byte[], byte[]> consumer = EasyMock.createStrictMock(Consumer.class);` then get rid of lines 637-643?

**Comment by bbejeck** [comment]
Left a couple of minor comments, assuming remaining comments from @mjsax are addressed, LGTM.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java (hunk 1 above)
Comment (kamalcph) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:325]: Refactored it.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java (hunk 2 above)
Comment (kamalcph) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:330]: Refactored it.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java (hunk 3 above)
Comment (kamalcph) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:113]: Earlier, It was logged as `trace` so left it as it is.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 4 above)
Comment (kamalcph) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:631]: Updated the code to use strict mock.

**Comment by kamalcph** [comment]
@mjsax @bbejeck 

Thanks for reviewing! I have done rebase with trunk but somehow the commits got polluted in this PR. So, opened a new [PR](https://github.com/apache/kafka/pull/4651) and closing this one.

**Pull request closed without merging**